File: mobile/flutter/lib/features/copy_trading/copy_trading_service.dart
// Copy Trading Service - Flutter Implementation

class Trader {
  final String id;
  final String username;
  final String avatar;
  final double winRate;
  final double totalPnL;
  final double pnlPercent;
  final int followers;
  final int copyCount;
  final String tradingPair;
  final double monthlyPnL;
  final double weeklyPnL;
  final double dailyPnL;
  final double maxDrawdown;
  final String avgHoldingTime;
  final String riskLevel;
  final bool isFollowing;
  final bool isPreInstalled;

  Trader({
    required this.id,
    required this.username,
    required this.avatar,
    required this.winRate,
    required this.totalPnL,
    required this.pnlPercent,
    required this.followers,
    required this.copyCount,
    required this.tradingPair,
    required this.monthlyPnL,
    required this.weeklyPnL,
    required this.dailyPnL,
    required this.maxDrawdown,
    required this.avgHoldingTime,
    required this.riskLevel,
    required this.isFollowing,
    required this.isPreInstalled,
  });
}

class CopyPosition {
  final String id;
  final String traderId;
  final String traderName;
  final String userId;
  final String symbol;
  final String side;
  final double size;
  final double entryPrice;
  final double currentPrice;
  final double pnl;
  final double pnlPercent;
  final DateTime openTime;
  final String status;

  CopyPosition({
    required this.id,
    required this.traderId,
    required this.traderName,
    required this.userId,
    required this.symbol,
    required this.side,
    required this.size,
    required this.entryPrice,
    required this.currentPrice,
    required this.pnl,
    required this.pnlPercent,
    required this.openTime,
    required this.status,
  });
}

class CopySettings {
  final String userId;
  final double copyAmount;
  final int copyLeverage;
  final double stopLossPercent;
  final double takeProfitPercent;
  final bool autoCopy;

  CopySettings({
    required this.userId,
    required this.copyAmount,
    required this.copyLeverage,
    required this.stopLossPercent,
    required this.takeProfitPercent,
    required this.autoCopy,
  });
}

class CopyTradingService {
  static List<Trader> generateTraders() {
    final List<Trader> traders = [];
    
    // Pre-installed top traders
    final preInstalledTraders = [
      {'username': 'CryptoWhale', 'avatar': '🐋', 'winRate': 78.5, 'totalPnL': 125000.0, 'pnlPercent': 156.2, 'followers': 15234, 'copyCount': 4521, 'pair': 'BTC/USDT', 'monthly': 12.5, 'weekly': 3.2, 'daily': 0.8, 'drawdown': -8.5, 'time': '2h 30m', 'risk': 'medium'},
      {'username': 'DeFiMaster', 'avatar': '🎯', 'winRate': 82.3, 'totalPnL': 98500.0, 'pnlPercent': 142.8, 'followers': 12456, 'copyCount': 3890, 'pair': 'ETH/USDT', 'monthly': 15.2, 'weekly': 4.1, 'daily': 1.2, 'drawdown': -6.2, 'time': '4h 15m', 'risk': 'low'},
      {'username': 'AltSeason', 'avatar': '🚀', 'winRate': 71.2, 'totalPnL': 87000.0, 'pnlPercent': 198.5, 'followers': 8923, 'copyCount': 2156, 'pair': 'SOL/USDT', 'monthly': 22.5, 'weekly': 8.3, 'daily': 2.1, 'drawdown': -12.8, 'time': '1h 45m', 'risk': 'high'},
      {'username': 'GridTrader', 'avatar': '📊', 'winRate': 85.1, 'totalPnL': 67800.0, 'pnlPercent': 98.3, 'followers': 6543, 'copyCount': 1890, 'pair': 'BNB/USDT', 'monthly': 8.2, 'weekly': 2.1, 'daily': 0.5, 'drawdown': -4.2, 'time': '6h 20m', 'risk': 'low'},
      {'username': 'MomentumKing', 'avatar': '👑', 'winRate': 75.8, 'totalPnL': 54200.0, 'pnlPercent': 125.6, 'followers': 9876, 'copyCount': 2567, 'pair': 'DOGE/USDT', 'monthly': 18.5, 'weekly': 5.2, 'daily': 1.5, 'drawdown': -15.2, 'time': '0h 45m', 'risk': 'high'},
      {'username': 'SwingTrader', 'avatar': '🌊', 'winRate': 68.5, 'totalPnL': 42500.0, 'pnlPercent': 88.2, 'followers': 5432, 'copyCount': 1234, 'pair': 'XRP/USDT', 'monthly': 10.2, 'weekly': 2.8, 'daily': 0.3, 'drawdown': -9.5, 'time': '12h 30m', 'risk': 'medium'},
      {'username': 'BotMaster', 'avatar': '🤖', 'winRate': 88.2, 'totalPnL': 38900.0, 'pnlPercent': 72.5, 'followers': 4321, 'copyCount': 987, 'pair': 'AVAX/USDT', 'monthly': 6.8, 'weekly': 1.5, 'daily': 0.2, 'drawdown': -3.2, 'time': '8h 00m', 'risk': 'low'},
      {'username': 'NanoGainer', 'avatar': '💎', 'winRate': 73.2, 'totalPnL': 31500.0, 'pnlPercent': 145.8, 'followers': 7654, 'copyCount': 1876, 'pair': 'PEPE/USDT', 'monthly': 25.2, 'weekly': 9.5, 'daily': 3.2, 'drawdown': -18.5, 'time': '0h 30m', 'risk': 'high'},
      {'username': 'StableTrader', 'avatar': '🛡️', 'winRate': 91.2, 'totalPnL': 28900.0, 'pnlPercent': 52.3, 'followers': 3210, 'copyCount': 654, 'pair': 'LINK/USDT', 'monthly': 4.2, 'weekly': 1.0, 'daily': 0.1, 'drawdown': -2.1, 'time': '24h 00m', 'risk': 'low'},
      {'username': 'FlashBoys', 'avatar': '⚡', 'winRate': 76.8, 'totalPnL': 24500.0, 'pnlPercent': 168.5, 'followers': 5678, 'copyCount': 1432, 'pair': 'MATIC/USDT', 'monthly': 14.5, 'weekly': 4.8, 'daily': 1.8, 'drawdown': -11.2, 'time': '1h 15m', 'risk': 'high'},
    ];

    for (int i = 0; i < preInstalledTraders.length; i++) {
      final t = preInstalledTraders[i];
      traders.add(Trader(
        id: 'trader-${i + 1}',
        username: t['username'] as String,
        avatar: t['avatar'] as String,
        winRate: t['winRate'] as double,
        totalPnL: t['totalPnL'] as double,
        pnlPercent: t['pnlPercent'] as double,
        followers: t['followers'] as int,
        copyCount: t['copyCount'] as int,
        tradingPair: t['pair'] as String,
        monthlyPnL: t['monthly'] as double,
        weeklyPnL: t['weekly'] as double,
        dailyPnL: t['daily'] as double,
        maxDrawdown: t['drawdown'] as double,
        avgHoldingTime: t['time'] as String,
        riskLevel: t['risk'] as String,
        isFollowing: false,
        isPreInstalled: true,
      ));
    }

    // Generate additional traders to reach 500+
    final avatars = ['🐵', '🦊', '🦁', '🐯', '🐲', '🐍', '🐴', '🦄', '🐝', '🦋'];
    final pairs = ['BTC/USDT', 'ETH/USDT', 'BNB/USDT', 'SOL/USDT', 'XRP/USDT'];
    final risks = ['low', 'medium', 'high'];

    for (int i = 0; i < 500; i++) {
      traders.add(Trader(
        id: 'trader-${i + 100}',
        username: 'Trader${i + 100}',
        avatar: avatars[i % avatars.length],
        winRate: 60 + (i % 30).toDouble(),
        totalPnL: 1000 + (i * 200).toDouble(),
        pnlPercent: 20 + (i % 200).toDouble(),
        followers: 100 + (i * 20),
        copyCount: 50 + (i * 10),
        tradingPair: pairs[i % pairs.length],
        monthlyPnL: (i % 30 - 5).toDouble(),
        weeklyPnL: (i % 10 - 2).toDouble(),
        dailyPnL: (i % 3 - 1).toDouble(),
        maxDrawdown: -(2 + (i % 20)).toDouble(),
        avgHoldingTime: '${i % 24}h ${i % 60}m',
        riskLevel: risks[i % 3],
        isFollowing: false,
        isPreInstalled: false,
      ));
    }

    return traders;
  }

  static List<Trader> getTopTraders() {
    return generateTraders().where((t) => t.isPreInstalled).toList();
  }

  static List<Trader> filterByRisk(List<Trader> traders, String risk) {
    if (risk == 'all') return traders;
    return traders.where((t) => t.riskLevel == risk).toList();
  }

  static List<Trader> sortByPnL(List<Trader> traders) {
    final sorted = List<Trader>.from(traders);
    sorted.sort((a, b) => b.totalPnL.compareTo(a.totalPnL));
    return sorted;
  }
}

File: backend/go/copy_trading_service/copy_trading_service.go
package copy_trading_service

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

// ============================================================================
// Types
// ============================================================================

type Trader struct {
	ID               string   `json:"id"`
	Username         string   `json:"username"`
	Avatar           string   `json:"avatar"`
	WinRate          float64  `json:"winRate"`
	TotalPnL         float64  `json:"totalPnL"`
	PnLPercent       float64  `json:"pnlPercent"`
	Followers        int      `json:"followers"`
	CopyCount        int      `json:"copyCount"`
	TradingPair      string   `json:"tradingPair"`
	MonthlyPnL       float64  `json:"monthlyPnL"`
	WeeklyPnL        float64  `json:"weeklyPnL"`
	DailyPnL         float64  `json:"dailyPnL"`
	MaxDrawdown      float64  `json:"maxDrawdown"`
	AvgHoldingTime   string   `json:"avgHoldingTime"`
	RiskLevel        string   `json:"riskLevel"`
	IsFollowing      bool     `json:"isFollowing"`
	IsPreInstalled   bool     `json:"isPreInstalled"`
	TotalTrades     int      `json:"totalTrades"`
	ProfitableTrades int      `json:"profitableTrades"`
}

type CopyPosition struct {
	ID           string    `json:"id"`
	TraderID     string    `json:"traderId"`
	TraderName   string    `json:"traderName"`
	UserID       string    `json:"userId"`
	Symbol       string    `json:"symbol"`
	Side         string    `json:"side"`
	Size         float64   `json:"size"`
	EntryPrice   float64   `json:"entryPrice"`
	CurrentPrice float64   `json:"currentPrice"`
	PnL          float64   `json:"pnl"`
	PnLPercent   float64   `json:"pnlPercent"`
	OpenTime     time.Time `json:"openTime"`
	Status       string    `json:"status"`
}

type CopySettings struct {
	UserID          string  `json:"userId"`
	CopyAmount      float64 `json:"copyAmount"`
	CopyLeverage    int     `json:"copyLeverage"`
	StopLossPercent float64 `json:"stopLossPercent"`
	TakeProfitPercent float64 `json:"takeProfitPercent"`
	AutoCopy        bool    `json:"autoCopy"`
}

type CopyTrade struct {
	ID          string    `json:"id"`
	TraderID    string    `json:"traderId"`
	UserID      string    `json:"userId"`
	Symbol      string    `json:"symbol"`
	Side        string    `json:"side"`
	Size        float64   `json:"size"`
	Price       float64   `json:"price"`
	Status      string    `json:"status"`
	CreateTime  time.Time `json:"createTime"`
}

// ============================================================================
// Service
// ============================================================================

type CopyTradingService struct {
	mu           sync.RWMutex
	traders      map[string]*Trader
	positions    map[string][]CopyPosition
	settings     map[string]*CopySettings
	trades       map[string][]CopyTrade
	followers    map[string][]string // traderID -> userIDs
}

func NewCopyTradingService() *CopyTradingService {
	cts := &CopyTradingService{
		traders:   make(map[string]*Trader),
		positions: make(map[string][]CopyPosition),
		settings:  make(map[string]*CopySettings),
		trades:    make(map[string][]CopyTrade),
		followers: make(map[string][]string),
	}
	cts.initializeDefaultTraders()
	return cts
}

func (cts *CopyTradingService) initializeDefaultTraders() {
	// Pre-installed top traders
	preInstalledTraders := []*Trader{
		{ID: "1", Username: "CryptoWhale", Avatar: "🐋", WinRate: 78.5, TotalPnL: 125000, PnLPercent: 156.2, Followers: 15234, CopyCount: 4521, TradingPair: "BTC/USDT", MonthlyPnL: 12.5, WeeklyPnL: 3.2, DailyPnL: 0.8, MaxDrawdown: -8.5, AvgHoldingTime: "2h 30m", RiskLevel: "medium", IsPreInstalled: true},
		{ID: "2", Username: "DeFiMaster", Avatar: "🎯", WinRate: 82.3, TotalPnL: 98500, PnLPercent: 142.8, Followers: 12456, CopyCount: 3890, TradingPair: "ETH/USDT", MonthlyPnL: 15.2, WeeklyPnL: 4.1, DailyPnL: 1.2, MaxDrawdown: -6.2, AvgHoldingTime: "4h 15m", RiskLevel: "low", IsPreInstalled: true},
		{ID: "3", Username: "AltSeason", Avatar: "🚀", WinRate: 71.2, TotalPnL: 87000, PnLPercent: 198.5, Followers: 8923, CopyCount: 2156, TradingPair: "SOL/USDT", MonthlyPnL: 22.5, WeeklyPnL: 8.3, DailyPnL: 2.1, MaxDrawdown: -12.8, AvgHoldingTime: "1h 45m", RiskLevel: "high", IsPreInstalled: true},
		{ID: "4", Username: "GridTrader", Avatar: "📊", WinRate: 85.1, TotalPnL: 67800, PnLPercent: 98.3, Followers: 6543, CopyCount: 1890, TradingPair: "BNB/USDT", MonthlyPnL: 8.2, WeeklyPnL: 2.1, DailyPnL: 0.5, MaxDrawdown: -4.2, AvgHoldingTime: "6h 20m", RiskLevel: "low", IsPreInstalled: true},
		{ID: "5", Username: "MomentumKing", Avatar: "👑", WinRate: 75.8, TotalPnL: 54200, PnLPercent: 125.6, Followers: 9876, CopyCount: 2567, TradingPair: "DOGE/USDT", MonthlyPnL: 18.5, WeeklyPnL: 5.2, DailyPnL: 1.5, MaxDrawdown: -15.2, AvgHoldingTime: "0h 45m", RiskLevel: "high", IsPreInstalled: true},
		{ID: "6", Username: "SwingTrader", Avatar: "🌊", WinRate: 68.5, TotalPnL: 42500, PnLPercent: 88.2, Followers: 5432, CopyCount: 1234, TradingPair: "XRP/USDT", MonthlyPnL: 10.2, WeeklyPnL: 2.8, DailyPnL: 0.3, MaxDrawdown: -9.5, AvgHoldingTime: "12h 30m", RiskLevel: "medium", IsPreInstalled: true},
		{ID: "7", Username: "BotMaster", Avatar: "🤖", WinRate: 88.2, TotalPnL: 38900, PnLPercent: 72.5, Followers: 4321, CopyCount: 987, TradingPair: "AVAX/USDT", MonthlyPnL: 6.8, WeeklyPnL: 1.5, DailyPnL: 0.2, MaxDrawdown: -3.2, AvgHoldingTime: "8h 00m", RiskLevel: "low", IsPreInstalled: true},
		{ID: "8", Username: "NanoGainer", Avatar: "💎", WinRate: 73.2, TotalPnL: 31500, PnLPercent: 145.8, Followers: 7654, CopyCount: 1876, TradingPair: "PEPE/USDT", MonthlyPnL: 25.2, WeeklyPnL: 9.5, DailyPnL: 3.2, MaxDrawdown: -18.5, AvgHoldingTime: "0h 30m", RiskLevel: "high", IsPreInstalled: true},
		{ID: "9", Username: "StableTrader", Avatar: "🛡️", WinRate: 91.2, TotalPnL: 28900, PnLPercent: 52.3, Followers: 3210, CopyCount: 654, TradingPair: "LINK/USDT", MonthlyPnL: 4.2, WeeklyPnL: 1.0, DailyPnL: 0.1, MaxDrawdown: -2.1, AvgHoldingTime: "24h 00m", RiskLevel: "low", IsPreInstalled: true},
		{ID: "10", Username: "FlashBoys", Avatar: "⚡", WinRate: 76.8, TotalPnL: 24500, PnLPercent: 168.5, Followers: 5678, CopyCount: 1432, TradingPair: "MATIC/USDT", MonthlyPnL: 14.5, WeeklyPnL: 4.8, DailyPnL: 1.8, MaxDrawdown: -11.2, AvgHoldingTime: "1h 15m", RiskLevel: "high", IsPreInstalled: true},
	}

	for _, trader := range preInstalledTraders {
		cts.traders[trader.ID] = trader
	}

	// Generate additional traders
	avatars := []string{"🐵", "🦊", "🦁", "🐯", "🐲", "🐍", "🐴", "🦄", "🐝", "🦋", "🌸", "🌺", "🌻", "🌹", "🍀"}
	pairs := []string{"BTC/USDT", "ETH/USDT", "BNB/USDT", "SOL/USDT", "XRP/USDT", "DOGE/USDT", "ADA/USDT", "AVAX/USDT", "DOT/USDT", "LINK/USDT"}
	riskLevels := []string{"low", "medium", "high"}

	for i := 0; i < 500; i++ {
		trader := &Trader{
			ID:             fmt.Sprintf("trader-%d", i+100),
			Username:       fmt.Sprintf("Trader%d", i+100),
			Avatar:         avatars[i%len(avatars)],
			WinRate:        60 + float64(i%30),
			TotalPnL:       1000 + float64(i*200),
			PnLPercent:     20 + float64(i%200),
			Followers:      100 + i*20,
			CopyCount:      50 + i*10,
			TradingPair:    pairs[i%len(pairs)],
			MonthlyPnL:     float64(i%30) - 5,
			WeeklyPnL:      float64(i%10) - 2,
			DailyPnL:       float64(i%3) - 1,
			MaxDrawdown:    -float64(2 + i%20),
			AvgHoldingTime: fmt.Sprintf("%dh %dm", i%24, i%60),
			RiskLevel:      riskLevels[i%3],
			IsFollowing:    false,
			IsPreInstalled: false,
		}
		cts.traders[trader.ID] = trader
	}
}

func (cts *CopyTradingService) GetAllTraders() []*Trader {
	cts.mu.RLock()
	defer cts.mu.RUnlock()

	traders := make([]*Trader, 0, len(cts.traders))
	for _, trader := range cts.traders {
		traders = append(traders, trader)
	}
	return traders
}

func (cts *CopyTradingService) GetTrader(traderID string) (*Trader, error) {
	cts.mu.RLock()
	defer cts.mu.RUnlock()

	trader, ok := cts.traders[traderID]
	if !ok {
		return nil, fmt.Errorf("trader not found: %s", traderID)
	}
	return trader, nil
}

func (cts *CopyTradingService) FollowTrader(userID, traderID string) error {
	cts.mu.Lock()
	defer cts.mu.Unlock()

	trader, ok := cts.traders[traderID]
	if !ok {
		return fmt.Errorf("trader not found: %s", traderID)
	}

	trader.IsFollowing = !trader.IsFollowing
	if trader.IsFollowing {
		trader.Followers++
		cts.followers[traderID] = append(cts.followers[traderID], userID)
	} else {
		trader.Followers--
		// Remove user from followers
		for i, id := range cts.followers[traderID] {
			if id == userID {
				cts.followers[traderID] = append(cts.followers[traderID][:i], cts.followers[traderID][i+1:]...)
				break
			}
		}
	}
	return nil
}

func (cts *CopyTradingService) GetUserFollowing(userID string) []*Trader {
	cts.mu.RLock()
	defer cts.mu.RUnlock()

	var following []*Trader
	for _, trader := range cts.traders {
		if trader.IsFollowing {
			following = append(following, trader)
		}
	}
	return following
}

func (cts *CopyTradingService) GetUserPositions(userID string) []CopyPosition {
	cts.mu.RLock()
	defer cts.mu.RUnlock()
	return cts.positions[userID]
}

func (cts *CopyTradingService) GetUserSettings(userID string) *CopySettings {
	cts.mu.RLock()
	defer cts.mu.RUnlock()

	if settings, ok := cts.settings[userID]; ok {
		return settings
	}
	return &CopySettings{
		UserID:          userID,
		CopyAmount:      1000,
		CopyLeverage:    1,
		StopLossPercent: 10,
		TakeProfitPercent: 20,
		AutoCopy:        false,
	}
}

func (cts *CopyTradingService) UpdateUserSettings(settings *CopySettings) error {
	cts.mu.Lock()
	defer cts.mu.Unlock()
	cts.settings[settings.UserID] = settings
	return nil
}

func (cts *CopyTradingService) CopyTrade(traderID, userID, symbol, side string, size float64) (*CopyPosition, error) {
	cts.mu.Lock()
	defer cts.mu.Unlock()

	trader, ok := cts.traders[traderID]
	if !ok {
		return nil, fmt.Errorf("trader not found: %s", traderID)
	}

	settings := cts.GetUserSettings(userID)
	price := 100.0 // Simulated price

	position := &CopyPosition{
		ID:           fmt.Sprintf("copy-pos-%d", time.Now().Unix()),
		TraderID:     traderID,
		TraderName:   trader.Username,
		UserID:       userID,
		Symbol:       symbol,
		Side:         side,
		Size:         size,
		EntryPrice:   price,
		CurrentPrice: price,
		PnL:          0,
		PnLPercent:   0,
		OpenTime:     time.Now(),
		Status:       "open",
	}

	cts.positions[userID] = append(cts.positions[userID], *position)
	return position, nil
}

func (cts *CopyTradingService) ToJSON() (string, error) {
	cts.mu.RLock()
	defer cts.mu.RUnlock()

	data := struct {
		Traders   map[string]*Trader   `json:"traders"`
		Positions map[string][]CopyPosition `json:"positions"`
		Settings  map[string]*CopySettings `json:"settings"`
	}{
		Traders:   cts.traders,
		Positions: cts.positions,
		Settings:  cts.settings,
	}

	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "", err
	}
	return string(jsonData), nil
}
